packages/yarnpkg-core/sources/Project.ts

```diff
@@ -703,9 +699,166 @@ export class Project {
 
       originalPackages.set(originalPkg.locatorHash, originalPkg);
 
-      const pkg = this.configuration.normalizePackage(originalPkg);
+      let pkg = this.configuration.normalizePackage(originalPkg);
+
+      const prettyParent = () => {
+        if (pkgParent)
+          return structUtils.prettyLocator(this.configuration, pkgParent);
+
+        const workspaceName = workspace.manifest.name;
+
+        if (workspaceName)
+          return structUtils.prettyIdent(this.configuration, workspaceName);
+
+        return `workspace without name`;
+      };
+
+      if (pkg.variants) {
+        // If a package has variants, remove the descriptorResolutionPromises so that it gets re-resolved each time it's a dependency,
+        // since its resolution might be different each time.
+        descriptorResolutionPromises.delete(structUtils.convertLocatorToDescriptor(pkg).descriptorHash);
+        packageResolutionPromises.delete(locator.locatorHash);
+        opts.report.reportInfo(MessageName.UNNAMED, `detected variant dependency, ${prettyParent()}'s dependency ${structUtils.prettyLocator(this.configuration, pkg)}, not caching resolution`);
```

Comment:
If there's always a `pkgParent`, then no need for `prettyParent`

```suggestion
        opts.report.reportInfo(MessageName.UNNAMED, `detected variant dependency, ${structUtils.prettyLocator(this.configuration, pkgParent)}'s dependency ${structUtils.prettyLocator(this.configuration, pkg)}, not caching resolution`);
```